First draft of custom buffer changes with block class memory allocation

dsorber opened the pull request:
interface and support for single mapped buffers. Note that support for single mapped buffer is not yet complete however many cases should work correctly.

Commit:
First draft of custom buffer changes with block class memory allocation
interface and support for single mapped buffers. Note that support for
single mapped buffer is not yet complete however many cases should work
correctly.

Commit:
Fix formatting and a couple other minor issues.

Commit:
Fix lingering formatting issue.

Commit:
Fix bug where buffer owner pointer needs to be a proper shared_ptr do
that the owning block is not freed before the buffer's deleter function
is called. Also add a DEBUG_SINGLE_MAPPED define to make it easier to
enable single mapped buffer debugging.

sdh11 commented:
I've been thinking more about the current take on the block API, and I think this approach is too limiting in terms of functionality. There are many cases where someone may need more fine-grained control over how and when memory transfers are scheduled that we really wouldn't want the block's work() to handle. I think we should use an inheritance based approach where this functionality would exist within the context of a buffer class and the block should only really need to focus on processing samples and not having to manage transfers (there could be caveats to that). If we put too much in work() for handling memory management then that begins to defeat the purpose of the blocks being more generic.

I completely agree with the sentiment of reducing the amount of boilerplate code someone would need to implement to use the custom buffers. It should still be possible to keep the boilerplate to a minimum with the inheritance based approach but maintain the ability for advanced users to override any functionality they need for their use-case. We can also have pre/post work/transfer hooks within the buffer class that could be implemented if needed.

I'm not sure what the arguments would look like, but at the moment, but the most simple case might look something like this:

```
// Custom buffer class
class cuda_pinned : public buffer_single_mapped {
  public:
    // Required override in buffer class
    char* allocate_buffer(size_t size) {
        ...
    }
}

class example_blk : virtual public sync_block {
  public:
    ...
};

class example_block_impl : public example_blk {
  public:
    buffer_sptr allocate_custom_buffer(...) {
        return gnuradio::make_buffer_sptr<cuda_pinned>(...);
    }
}
```

sdh11 commented:
We might also want to use the same kind of inheritance structure for the buffer_readers that you do for the buffers just for consistency (buffer_reader_sm/buffer_reader_dm?).

Commit:
Modify single mapped buffer implementation to use explicit
input_blocked_callback and output_blocked_callback functions when input
or output (respectively) are blocked.

Commit:
Fix a couple minor issues from previous commit including missing throw
statement and errant debug message.

Commit:
Fix another issue; the items_avail value passed into the
input_blocked_callback() might be stale so recalculate items_avail if
the body of the callback is executed.

Commit:
Fix issue where block is "effectively output blocked" when its call to
general_work() produces no output by calling the output blocked callback in
this case.

Commit:
work-in-progress

Commit:
Add callback ready functions to slightly improve performance in some
cases.

mormj commented on gnuradio-runtime/lib/block_executor.cc:
```diff
@@ -260,17 +289,40 @@ block_executor::state block_executor::run_one_iteration()
+            if (out_buf->output_blkd_cb_ready(m->output_multiple())) {
+                gr::custom_lock lock(std::ref(*out_buf->mutex()), out_buf);
+                if (!out_buf->output_blocked_callback(m->output_multiple(),
+                                                      m->min_noutput_items())) {
```
Why is the second argument a bool, but int (min_noutput_items()) is input to the output_blocked_callback function

dsorber replied:
Whoops... the min number of outputs items was passed in as a parameter to the callback at one point but was later found to not be needed and removed.  Somehow I missed this during cleanup. An update will be forthcoming.

mormj commented on gnuradio-runtime/lib/buffer_reader_sm.cc:
```diff
@@ -0,0 +1,149 @@
+        GR_LOG_DEBUG(d_logger, msg.str());
+#endif
+
+        // Shift existing data down to make room for blocked data at end of buffer
```
I'm wondering if a lot of this logic should be moved into the buffer class since it uses all the private variables

Commit:
Remove legacy parameters from output_blocked_callback() calls.

dsorber committed:
Merge branch 'master' into mem-alloc-inf-dev

Commit:
Correct header file hashes for pybind.

Commit:
Fix miscellaneous warnings.

Commit:
Fix minor issues with portaudio_sink and portaudio_source blocks.

Commit:
Adding make_buffer_double_mapped() function to fix the qa_buffer tests
which are now essentially unit tests for the buffer_double_mapped class.
